### ISSUE TITLE
testCase / resultNode name and type

### ISSUE THREAD

**Issue by StrayAlien** [dmn-tck/tck#348]
Hello all,

After a fair bit of 'life happens' this year I am getting our code base back up to date with the TCK.  I am going through the interim TCK modifications and have some questions. 

This first involves the changes to the way the test suite determines what to execute, and how to interpret the results from that execution.  I am referring to the change of `type` from resultNode to testCase and the resulting overloading of resultNode `name`.   Correct me if I am wrong here:

When the `type` of a testCase is anything but `'decisonService'`, the testcase `invocableName` is ignored and the name of the DMN 'thing' to assert on comes from the resultNode name.  So, decision test name comes from the resultNode name.  

But, when the testCase type is `'decisionService'`, then the name of the DMN 'thing' to assert on comes from the testCase `invocableName` (not from the resultNode `name`) and the resultNode `name` is now something different entirely - I believe now is meant to be a single context item name of the decision service result (?).

So, depending on what 'thing' we're testing, its name comes from different places, and we interpret the result differently.

If the above is accurate, then besides the confusing nature of type/name etc here, it also means the TCK Decision Service tests no longer assert the _complete_ result of a decision service invocation - just the one item named on the resultNode - the decision service tests have been changed as such.  I believe we should assert on the total result of the decision service innvocation, not some selective portion of it.

To my mind the way type/name worked before was pretty easy, in fact, we didn't even 'type' on neither testCase nor resultNode, just the name on the resultNode to figure what DMN thing we're asserting on.  In, fact, I thought `type` there was just documentation ...

I can't recall thinking there was anything wrong with the way `name` on resultNode worked before. Was there something actually wrong with it? Was the TCK changed to match a particular DMN implementation?

As a suggestion.  Have just 'name' on either testCase or resultNode (but not both) and lose the notion of 'type' entirely.  With name on testCase, you restrict the testCase to a single invocation of decision/DS/etc (and should also restrict the testCase to a single resultNode).  Alternatively, with it on resultNode you can make multiple DMN invocations / assertions per testCase.  

Greg.

**Comment by tarilabs** [comment]
Hi Greg,
nice to hear back from you, and I hope things are well (/better now?).

I don't believe there was a "before"/after.
TCK originally did not actually support decisionservice test.
We welcomed your contribution of decisionservice test cases, in fact thank you again for pushing the envelope forward.

In order to support from TCK infrastructure point of view execution of decisionservice (and in the future BKM in isolation) we introduced the supporting infrastructural changes with https://github.com/dmn-tck/tck/pull/308
I tried my best there as a tck maintainer to link all related conversation and changes. So you can check there what was discussed and approved.

&nbsp;

I believe your statements need some additional notes or correction, below.

> When the type of a testCase is anything but 'decisonService', the testcase invocableName is ignored and the name of the DMN 'thing' to assert on comes from the resultNode name. So, decision test name comes from the resultNode name.

Correct. You can also have multiple resultNode(s) for the test case. This is the scenario when is meant to test by execution of the whole model. Implementor are free to choose if execute the whole model and assert for each result node, or invoke only the asserted result node DMN-Decisions. This is compatible with the deliberate vagueness of the DMN spec wrt the API call.

&nbsp;

> But, when the testCase type is 'decisionService', then the name of the DMN 'thing' to assert on comes from the testCase invocableName (not from the resultNode name) and the resultNode name is now something different entirely - I believe now is meant to be a single context item name of the decision service result (?).

Correct. But then again, if you had a decisionservice with multiple output decisions you can have a testcase with multiple resultnodes and that would be corresponding to the outputdecisions of the decisionservice.

&nbsp;

> Was the TCK changed to match a particular DMN implementation?

Do you have elements to support this claim, or is this just to deliberately provoke?
Because as I cannot find in your report which elements would support your claim, hence by exclusion I am led to infer this is not a constructive question the way it is.

On other hand we have had a multiple implementors agreement on https://github.com/dmn-tck/tck/pull/308 and multiple implementors submitted results for the same consensus.

May I suggest being this a community effort, if you want to have a discussion kindly frame the question in a way it can be proactively/constructively discussed?

&nbsp;

> With name on testCase, you restrict the testCase to a single invocation of decision/DS/etc

This is wrong.
When you invoke without specifying an invocable, you can assert on multiple decisions.

&nbsp;

> (and should also restrict the testCase to a single resultNode)

This is wrong. Even when you specify an invocable, you can assert multiple resultnodes. If the invocable is a decisionservice, they would correspond to the decisionservice's multiple outputdecision(s).
Indeed if the test case specify an invocable, the implementor is expected to invoke said invocable. For testing several invocables, specify multiple test cases.

&nbsp;

headsup: I am focusing my effort to align test cases to the upcoming DMNv1.3, and you can find an open PR with the work in progress. There are some tests which are going to be impacted by the DMNv1.3 changes.

Best regards,
MM

**Comment by StrayAlien** [comment]
Thanks @tarilabs.  Not looking to provoke. I wasn't claiming anything. 

I hope you and the crew are well too.  I've had quite some things / life to attend to, hopefully time and responsibilities and family health will permit a bit more time here.   

I have a feeling this one will be filed under "its already done and nobody is going to change it" but I'll give it a go anyways - things can always be better / simpler.

To my mind it feels overcomplicated.  The "this is wrong" responses above highlight that (and I was not stating something there - those were suggestions ... ).  Not only do I have to know where to get the name from I also have to know what to do when something does not have a value.  

Is there any particular reason why we need to distinguish what type of thing is being asserted on?  

IMO, I don't believe we need to distinguish when a test is for a decision service or a decision or a bkm.   Just name the thing you're asserting against, and assert on the result.  We don't have to be concerned with whether the impl runs one thing or everything.  Not important - let the test runner impl or the runtime impl decide.  Something needs to distinguish it, but I don't believe the test needs to know.  Like, good tests don't know about implementation details. In this case, if I change the impl from one drg type to another then the test has to change .....

Having just the name on the resultNode (I think) caters for all the scenarios you mentioned above.  You can assert on one or multiple things per testcase.     

Think of it like a restful call, one provides the input, makes a call and gets a result.  I am not concerned with how it is implemented.  

On top of that, I still believe the idea of asserting against individual context items for a decision service result is not good testing practice.  A decision service invocation returns a context, and as such we should assert that entire context is correct.  When we test a decision that returns a context we assert on the entire context- not just specifically names parts of it - why would we assert differently on a context just because it is returned by a decision service?

Even if the complexity of the name/invocableName/type thing remains, I believe we should at least keep context assertions consistent across decisions, decision services, and any other types we might test that return context.  Assert on the entire result.

Greg.   

PS.  Re the 1.3 thing, I saw that PR but as yet have not managed to find any draft spec for 1.3 or anything detailing the differences.  Is there such a draft available?  Feel free to answer on that PR to keep this focussed.

**Comment by tarilabs** [comment]
As you can notice in the original discussion, https://github.com/dmn-tck/tck/pull/308/files#diff-c8d71167e18a3144bd0b8ddd8d86eea7R8 the original design of the TCK was to state what is the scope of the test (decisions Vs BKM).
This dates back even before I myself joined the TCK tbc.
The TCK design was augmented, naturally, to support invocation and testing of decision service.
Multiple implementors agreed on this natural augmentation of the original design.

You are proposing a different design now.

As I stated in my reply, the claimed limitations of the current design you stated when opening this report are mostly partial views (in my reply: "_Correct._ [...]") or not correct (in my reply: "_This is wrong._ [...]").

You are correct when you state the current design support for each test case only 1 invocable Decision Service; as I also stated in my reply, nothing prevents you to have multiple test cases in the same file, and test each Decision Service in isolation which is a _good practice_ of unit testing.

I will leave to others to comment if they want to redesign the TCK infrastructure and change all their runners to support your alternative proposal.

**Comment by StrayAlien** [comment]
Hi @tarilabs - thanks again.  I'm not proposing a different design at all - how it was when I submitted the PRs was just fine - I don't think it needed changing is more the point.

When I submitted the PRs my runner was executing decision services (and in my own tests - BKMs) without any modifications to the test xsd.  So, clearly it is not actually required.  It was simple and worked.

Just saying I don't believe we need to distinguish at the test level what type of thing is being invoked.  It already worked and was already future proof and supported multiple DS / decision / BKM invocation combinations within a single testcase etc etc.  However, I can live with it.

Perhaps more importantly, I am also saying that we really should assert on the entire decision service context result in tests, not just a part of it.  This was how my PRs asserted them and I believe it is the correct thing to do.

The way it is now leads to this:

* if asserting the result of a decision that returns a context - we assert the entire context result
* if a decision gets its result by calling a decision service, we're asserting on an entire context result
* but ... if we call the decision service directly we're not asserting the entire context result.  We have to name just the parts we're asserting on.

It's like calling a decision service is a special thing - which I don't think it is.  It is a function, it returns a result - let's assert the result.  BKMs would be the same.

Greg.
